embulk-standards/src/test/java/org/embulk/standards/TestLocalFileInputPlugin.java

```diff
@@ -0,0 +1,173 @@
+package org.embulk.standards;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import com.google.common.base.Optional;
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import org.embulk.EmbulkTestRuntime;
+import org.embulk.config.ConfigSource;
+import org.embulk.spi.Exec;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+/**
+ * Tests LocalFileInputPlugin.
+ */
+public class TestLocalFileInputPlugin {
+    @Rule
+    public EmbulkTestRuntime runtime = new EmbulkTestRuntime();
+
+    @Rule
+    public TemporaryFolder workdir = new TemporaryFolder();
+
+    @Test
+    public void testListFiles() throws IOException {
+        final LocalFileInputPlugin.PluginTask task = buildTask("foofoo");
+        this.workdir.newFile("foofoo1");
+        this.workdir.newFile("FooFoo2");
+        this.workdir.newFile("barbar3");
+        this.workdir.newFolder("foofoo4");
+        this.workdir.newFile("foofoo4/foo");
+        this.workdir.newFile("foofoo4/bar");
+        this.workdir.newFolder("barbar5");
+        this.workdir.newFile("barbar5/foo");
+        this.workdir.newFile("barbar5/bar");
+        this.workdir.newFolder("fooFoo6");
+        this.workdir.newFile("fooFoo6/foo");
+        this.workdir.newFile("fooFoo6/bar");
+        final LocalFileInputPlugin plugin = new LocalFileInputPlugin();
+        final List<String> files = plugin.listFiles(task);
+
+        // It intentionally tests in the platform-aware way, not in the platform-oblivious way.
+        if (System.getProperty("os.name").contains("Windows")) {
+            assertEquals(6, files.size());
+            assertTrue(files.contains(buildPath("foofoo1")));
+            assertTrue(files.contains(buildPath("FooFoo2")));
+            assertTrue(files.contains(buildPath("foofoo4\\foo")));
+            assertTrue(files.contains(buildPath("foofoo4\\bar")));
+            assertTrue(files.contains(buildPath("fooFoo6\\foo")));
+            assertTrue(files.contains(buildPath("fooFoo6\\bar")));
+        } else if (System.getProperty("os.name").contains("Mac OS")) {
```

Comment:
I expect Mac OS will need the same care with Windows. (The default file system on Mac OS handles file names case-insensitively.)

Reply:
Hmm, however, the test at f6d7ef5 passed on my Mac enviroment.

```console
$ uname -a
Darwin baymax.local 17.6.0 Darwin Kernel Version 17.6.0: Tue May  8 15:22:16 PDT 2018; root:xnu-4570.61.1~1/RELEASE_X86_64 x86_64
```

`System.getProperty("os.name")` return the `Mac OS X`

```console
$ javac -version
javac 1.8.0_131
```

Reply:
@jca02266 Got it. Reproduced on my environment as well... But if so, it means that this fix is not perfect yet. It is still working inconsistently on Mac OS between the "directory name" part and the "base file name" part.

For the `testListFilesWithDifferentCaseDirectoryPrefix` case:
* `prefix_path:"Directory1/foo"` matched with `directory1/foo1` and `directory1/foo2`
* `prefix_path:"Directory1/foo"` didn't match with `directory1/Foo3`.

We at least need to confirm why this code is working like that before accepting/merging this change.

Reply:
Hmm, `FileSystems.getDefault().getPathMatcher().getClass().getName()` return the `sun.nio.fs.UnixFileSystem$3`.

`FileSystems.getDefault().getPathMatcher(INCORRECT_GLOB_PATTERN)` show following stacktrace:

```
Caused by: java.util.regex.PatternSyntaxException: Missing '] near index 7
[foofoo*
       ^
	at sun.nio.fs.Globs.toRegexPattern(Globs.java:146)
	at sun.nio.fs.Globs.toUnixRegexPattern(Globs.java:211)
	at sun.nio.fs.UnixFileSystem.getPathMatcher(UnixFileSystem.java:294)
...
```

Reply:
On the other hand, IIUC, Java on Mac OS counts case-insensitively-matched directories in `Files.walkFileTree`, right?

In summary on Mac OS (in my understanding):
* `Files.walkFileTree` includes directories with different cases
* `FileSystems.getDefault().getPathMatcher` does not include directories with different cases

Reply:
It looks coming from Java's internal implementation:
* `Files.walkFileTree`: (estimated to be) calling the operating system's tree walking API
* `PathMatcher`: matching processed in Java by building `java.util.regex.Pattern` in [`MacOSXFileSystem`](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/43cb25339b55/src/solaris/classes/sun/nio/fs/MacOSXFileSystem.java#l49) in a case-sensitive way

----

To make it at least "consistent", I'd suggest:

* `Files.walkFileTree` is still used in the same way for listing candidates.
* But, `PathMatcher` is applied to the entire path name, not just the base file name part.

By this way, the matching works at least consistent. Case-sensitive matching is applied in Mac OS X , though.

What do you think?

Reply:
I'm considering and trying to imprement your suggestion.
Please wait for a while.
At least, it may become complicated to escape glob meta character because '\' is meta and windows path separater.

Reply:
And If it complicate, I think that the current changes is best. At least, on Windows, it become well. On the other, on Mac, it is not changed. The case-sensitive behaviour of Mac is java's problem.

Reply:
> And If it complicate, I think that the current changes is best. At least, on Windows, it improved. On the other, on Mac, it is not changed. The case-sensitive behaviour of Mac is java's problem.

It'd be acceptable, and I don't think the behavior of the current implementation on Mac is Java's problem. It's Embulk's problem. The problem is because Embulk is using a combination of two different methods to identify files from a prefix string. After we use only `PathMatcher` to identify files, it'd be acceptable to say "It's Java's problem".

I don't think this will be a dangerous change for Mac users because it causes neither any partial success nor any unintended pick-ups. When this change affects, it fails at all. Users can find it out.

Reply:
real path: /foo
path_prefix: /Foo

In this case, at first, embulk open the directory "Foo" , The Path object has the name "Foo" instead of "foo".
I think We must only use the realpath() system call to get the case sensitive real path, however, realpath() will be resolve symbolic link. So we can't use realpath() because If /foo is linked to /bar, realpath() return the "/bar".

> PathMatcher is applied to the entire path name, not just the base file name part.

Simply, I think we can't do that.

```
public class Sample {
  @Rule
  public TemporaryFolder temp = new TemporaryFolder();

  @Test
  public void test() throws Exception {
    final Path parent = temp.getRoot().toPath();

    temp.newFolder("foo", "BAR");
    temp.newFile("foo/BAR/Baz");

    Files.createSymbolicLink(Paths.get(parent.toString(), "foo/qux"),
                             Paths.get(parent.toString(), "foo/BAR"));

    Path first = Paths.get(parent.toString(), "Foo");

    assertThat(first.toString(), endsWith("Foo"));
    assertThat(first.toRealPath().toString(), endsWith("foo"));
    // If it is passed LinkOption.NOFOLLOW_LINKS, realpath() is not called
    assertThat(first.toRealPath(LinkOption.NOFOLLOW_LINKS).toString(), endsWith("Foo")); // "foo" is expected, but "Foo"

    Path second = Paths.get(parent.toString(), "Foo/qux");

    assertThat(second.toString(), endsWith("Foo/qux"));
    // If Foo/qux is symbolic link, it is resolved.
    assertThat(second.toRealPath().toString(), endsWith("foo/BAR")); // "foo/qux" is expected, but "foo/BAR"
    assertThat(second.toRealPath(LinkOption.NOFOLLOW_LINKS).toString(), endsWith("Foo/qux")); // "foo/qux" is expected, but "Foo/qux"

    EnumSet<FileVisitOption> option = EnumSet.noneOf(FileVisitOption.class);
    int maxDepth = Integer.MAX_VALUE;

    Files.walkFileTree(first, option, maxDepth, new SimpleFileVisitor<Path>() {
      @Override
      public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
        System.out.println("path: " + file.toString());
        System.out.println("real: " + file.toRealPath().toString());
        return FileVisitResult.CONTINUE;
      }
    });
  }
}
```

Reply:
Sorry, I don't get your point... What do you want to point/claim?

Reply:
oh, sorry.

> Files.walkFileTree is still used in the same way for listing candidates.
> But, PathMatcher is applied to the entire path name, not just the base file name part.

I tried to implemente above, and I found that it is imposible. 

1. We must get the real path (real case)  before applying PathMatcher.
2. We can use the realpath() system call to get the real path.
3. However, realpath() might also resolve symbolic link to real path, so PathMatcher can never match such the path. 

Am I making sense?

If you can't make sense yet (probably, my English is bad), I will implement above code.

Reply:
@jca02266 Ah, okay... I eventually got your point after some experiments.

Finally, let us accept (merge) the current implementation of yours as the second best choice for the reason you mentioned :
> At least, on Windows, it improved. On the other, on Mac, it is not changed.

But let us consider a bit more whether we'd release this as the next release `0.9.8` or not.